File: Sort, Search, Tree1(BFT, DFT, BFS, DFS)/Sort/merge_sort.c
#include <stdio.h>
#include<stdlib.h> // malloc

void merge(int L[], int first, int mid, int last){
    /*
    L는 정렬할 대상
    */

    int k = first;
    int i = 0;
    int j = 0;
    int n1 = mid -first +1; // size of sub1
    int n2 = last - mid; // size of sub2

    //subarray
    int * sub1 = (int*)malloc(n1*sizeof(int));
    int * sub2 = (int*)malloc(n2*sizeof(int)); 
    // copying
    for(int i =0; i<n1; i++){
        sub1[i] = L[first  +i];
    }
    for(int j=0;j<n2; j++){
        sub2[j] = L[mid +1 +j];
    }

    //merge
    i = 0;
    j= 0;
    k = first;

    while(i<n1 && j<n2){
        if(sub1[i] <=sub2[j]){
            L[k] = sub1[i];
            i++;
        }
        else{
            L[k] = sub2[j];
            j++;
        }
        k++;
    }

    //handling remains
    while(i<n1){
        L[k] = sub1[i];
        i++;
        k++;
    }
    while(j<n2){
        L[k] = sub2[j];
        j++;
        k++;
    }

    //free dynamic arrays
    free(sub1);
    free(sub2);

    

}



void mergeSort(int L[], int first, int last){
    
    if(first<last){
        int mid = first + (last - first)/2;
        mergeSort(L, first, mid);
        mergeSort(L, mid+1, last);
        merge(L, first, mid, last);
    }
}





int main(void){

    int array[] = {7,8,5,8,3,5,9,1};
    int array_size =sizeof(array) / sizeof(array[0]);
    mergeSort(array, 0, array_size-1);

    for (int i= 0; i<array_size ; i++){
        printf("%d", array[i]);
    }
}

File: Sort, Search, Tree1(BFT, DFT, BFS, DFS)/Sort/merge_sort.py
'''
Merge Sort

1. recursion method
2. time complexity : O(N*logN)
3. 항상 반을 나누고 -> LogN, 합치는 순간 정렬한다. ->N
    3-1. 항상 recursively 끝까지 반을 나누다가 나중에 합치기.
4. "이미 정렬된 두 부분집합"을 합친다.

'''


def merge(L, first, mid, last): # 합치면서 정렬
    k = first

    #mid 나뉜 포인트 -> 이제부터 L로 다시 합치면서 정렬할 것임.
    sub1 = L[first:mid +1]
    sub2 = L[mid+1 : last+1]


    i =j =0 #sub array의 처음부터 검사.

    # 두 인덱스를 모두 순회하면서 어느 하나가 끝나기 전까지 순회하면서, 정렬
    while(i<len(sub1) and j< len(sub2)):
        if sub1[i] <=sub2[j]:
            L[k] = sub1[i]
            i+=1

        else:
            L[k] = sub2[j]
            j+=1

        k+=1 # 반드시 포함해야 하는 부분.
    
    #둘 중 남은 부분 처리하기
    if(i<len(sub1)):
        L[k:last+1] = sub1[i:]

    if(j<len(sub2)):
        L[k : last+1] = sub2[j:]


def Mergesort(L, first, last):
    '''
    L: 정렬할 대상
    first: 정렬할 대상의 정렬할 곳의 시작
    last: 정렬할 대상의 정렬할 곳의 마지막
    '''

    if first == last:
        return
    
    else:
        mid = first + ((last -first)//2) ##  mid = (first + last) // 2도 가능하지만 오버플로우를 방지하기 위함.

        #계속 오른쪽 왼쪽 나눠가면서, 작은쪽이 크기가 1에 달할때까지 나누고  그 뒤로 왼쪽부터 merge해 나간다.
        #다 끝나면 오른쪽의 subarray들을 merge
        Mergesort(L, first, mid)
        Mergesort(L, mid+1, last)

        merge(L, first, mid, last)

    
def main():
    L = [2,5,6,8,2,10]

    Mergesort(L, first = 0, last = len(L)-1)
    print(L)

if __name__ == "__main__":
    main()